llama_index/llm_predictor/base.py

```diff
@@ -193,6 +201,9 @@ def get_llm_metadata(self) -> LLMMetadata:
         else:
             return LLMMetadata()
 
+    def _get_formatted_prompt(self, prompt: Prompt, **prompt_args: Any) -> str:
+        return prompt.format(llm=self._llm, **prompt_args)
```

Comment:
Question 1: How critical is it to keep track of the formatted prompts that are used? Seems like they're only used for counting tokens?

Assuming it is critical, Vellum will need to hit an API to get the formatted prompt, hence moving this into a helper that can be overridden.

Reply:
Good question. AFAIK mostly used for counting tokens and logging

Reply:
Probably dumb question, couldn't we just format the string (locally, outside of Vellum service) and return that?

Reply:
> couldn't we just format the string

This _may_ work today, but I have concerns with this because:
1. We'd have to always keep an eye out in logical drift between how Vellum's backend compiles prompts and how LlamaIndex does
2. Vellum will soon support "function calls" within the Jinja template that require Vellum's backend to fully compile.

In short, Vellum Prompt Templates aren't quite as simple as string substitutions.